airbyte-integrations/connectors/destination-clickhouse/src/main/kotlin/io/airbyte/integrations/destination/clickhouse/schema/ClickhouseTableSchemaMapper.kt

```diff
@@ -55,30 +57,66 @@ class ClickhouseTableSchemaMapper(
         // Map Airbyte field types to ClickHouse column types
         val clickhouseType =
             when (fieldType.type) {
-                BooleanType -> "Bool"
-                DateType -> "Date32"
-                IntegerType -> "Int64"
-                NumberType -> DECIMAL_WITH_PRECISION_AND_SCALE
-                StringType -> "String"
-                TimeTypeWithTimezone -> "String"
-                TimeTypeWithoutTimezone -> "String"
+                BooleanType -> ClickhouseSqlTypes.BOOL
+                DateType -> ClickhouseSqlTypes.DATE32
+                IntegerType -> ClickhouseSqlTypes.INT64
+                NumberType -> ClickhouseSqlTypes.DECIMAL_WITH_PRECISION_AND_SCALE
+                StringType -> ClickhouseSqlTypes.STRING
+                TimeTypeWithTimezone -> ClickhouseSqlTypes.STRING
+                TimeTypeWithoutTimezone -> ClickhouseSqlTypes.STRING
                 TimestampTypeWithTimezone,
-                TimestampTypeWithoutTimezone -> DATETIME_WITH_PRECISION
+                TimestampTypeWithoutTimezone -> ClickhouseSqlTypes.DATETIME_WITH_PRECISION
                 is ArrayType,
                 ArrayTypeWithoutSchema,
                 is UnionType,
-                is UnknownType -> "String"
+                is UnknownType -> ClickhouseSqlTypes.STRING
                 ObjectTypeWithEmptySchema,
                 ObjectTypeWithoutSchema,
                 is ObjectType -> {
                     if (config.enableJson) {
-                        "JSON"
+                        ClickhouseSqlTypes.JSON
                     } else {
-                        "String"
+                        ClickhouseSqlTypes.STRING
                     }
                 }
             }
 
         return ColumnType(clickhouseType, fieldType.nullable)
     }
+
+    override fun toFinalSchema(tableSchema: StreamTableSchema): StreamTableSchema {
+        if (tableSchema.importType !is Dedupe) {
+            return tableSchema
+        }
+
+        // For dedupe mode we do extra logic to ensure certain columns are non-null:
+        //     1) the primary key columns
+        //     2) the version column used by the dedupe engine (in practice the cursor)
+        val pks = tableSchema.getPrimaryKey().flatten()
```

Comment:
do we check somewhere else that all PK entries are length 1? otherwise this will turn `[[foo, bar]]` into `[foo, bar]` which seems wrong